torch/testing/_internal/distributed/rpc/rpc_test.py

```diff
@@ -2582,6 +2583,28 @@ def test_callback_none(self):
         ):
             rpc.rpc_async(dst, raise_func)._then(None)
 
+    @dist_init
+    def test_rref_timeout(self):
+        # This test is similar to ones in FaultyProcessGroupTest, but is meant to be
+        # run with other backends besides ProcessGroup.
+        if self.rank != 0:
+            return
+
+        dst_rank = (self.rank + 1) % self.world_size
+        dst_worker = "worker{}".format(dst_rank)
+        # 10 ms timeout
+        rref = rpc.remote(dst_worker, my_sleep_func, args=(2, ), timeout=0.01)
+        # Future corresponding to the remote creation should time out.
+        expected_error = get_timeout_error_regex(dist_utils.TEST_CONFIG.rpc_backend_name)
+        with self.assertRaisesRegex(RuntimeError, expected_error):
+            rref._get_future().wait()
+        # Call to ensure pending callbacks are run.
+        wait_until_pending_futures_and_users_flushed()
+        with self.assertRaisesRegex(RuntimeError, "RRef creation"):
+            rref.to_here()
+
+        wait_until_owners_and_forks_on_rank(1, 1, rank=1)
```

Comment:
does this mean applications also need to call this to wait for deletion?

Reply:
No, this isn't expected to be called by users, it is for tests where we force specific errors (i.e. a timeout error on the client, but the server eventually processes the OwnerRRef creation, for example). Waiting in this way in tests will prevent flakiness/timeouts. In general, the user application won't really have a way to distinguish whether a message was completely dropped, or has timed out but will be completed at some point in the future. 

If the latter case is true in the user application, then there is a chance that RPC on the client node will shut down, and in this case the `RRefUserDeletes` sent out by the client won't have any effect. And if the server node shuts down after (in particular runs `_delete_all_user_rrefs`) it will incur a blocking timed wait on the owners and forks map in rref_context. This won't throw errors, but will delay RPC shutdown. I think if this becomes an issue, we can expose knobs to the user to control how long to wait on these messages during shutdown.